skcosmo/feature_selection/_greedy.py

```diff
@@ -0,0 +1,225 @@
+            run, after increasing n_features_to_select.
+            Assumes it is called with the same X and y
+
+        Returns
+        -------
+        self : object
+        """
+        tags = self._get_tags()
+
+        if y is not None:
+            X, y = self._validate_data(
+                X,
+                y,
+                accept_sparse="csc",
+                ensure_min_features=2,
+                force_all_finite=not tags.get("allow_nan", True),
+                multi_output=True,
+            )
+        else:
+            X = check_array(
+                X,
+                accept_sparse="csc",
+                ensure_min_features=2,
+                force_all_finite=not tags.get("allow_nan", True),
+            )
+
+        n_features = X.shape[1]
+
+        error_msg = (
+            "n_features_to_select must be either None, an "
+            "integer in [1, n_features - 1] "
+            "representing the absolute "
+            "number of features, or a float in (0, 1] "
+            "representing a percentage of features to "
+            f"select. Got {self.n_features_to_select} features and "
+            f"an input with {n_features} features."
+        )
+
+        if self.n_features_to_select is None:
+            n_iterations = n_features // 2
+        elif isinstance(self.n_features_to_select, numbers.Integral):
+            if not 0 < self.n_features_to_select < n_features:
+                raise ValueError(error_msg)
+            n_iterations = self.n_features_to_select
+        elif isinstance(self.n_features_to_select, numbers.Real):
+            if not 0 < self.n_features_to_select <= 1:
+                raise ValueError(error_msg)
+            n_iterations = int(n_features * self.n_features_to_select)
+        else:
+            raise ValueError(error_msg)
+
+        if warm_start:
+            if self.n_selected_ == 0:
+                raise ValueError(
+                    "Cannot fit with warm_start=True without having been previously initialized"
+                )
+            self._continue_greedy_search(X, y, n_iterations)
+        else:
+            self.n_selected_ = 0
+            self._init_greedy_search(X, y, n_iterations)
+
+        n_iterations -= self.n_selected_
+
+        for n in self.report_progress(range(n_iterations)):
+
+            new_feature_idx = self._get_best_new_feature(self.scoring, X, y)
+            if new_feature_idx is not None:
+                self._update_post_selection(X, y, new_feature_idx)
+                self._postprocess()
+            else:
+                warnings.warn(
+                    f"Score threshold of {self.score_threshold} reached."
+                    f"Terminating search at {self.n_selected_} / {self.n_features_to_select}."
+                )
+                self.X_selected_ = self.X_selected_[:, :n]
+                self.selected_idx_ = self.selected_idx_[:n]
+                break
```

Comment:
Stylistic question - is `break` correct to use? I have memorised as a mantra that it is highly undesirable.

Reply:
ha, that's very true -- I'll update.